doc/neps/nep-0050-scalar-promotion.rst

```diff
@@ -0,0 +1,729 @@
+
+.. note::
+
+    See alternatives, we may yet decide that this silent overflow is acceptable
+    or at least a separate issue.
+
+
+Alternatives
+============
+
+There are several design axes where different choices are possible.
+The below sections outline these.
+
+Use strongly-typed scalars or a mix of both
+-------------------------------------------
+
+The simplest solution to the value-based promotion/casting issue would be to use
+strongly typed Python scalars, i.e. Python floats are considered double precision
+and Python integers are always considered the same as the default integer dtype.
+
+This would be the simplest solution, however, it would lead to many upcasts when
+working with arrays of ``float32`` or ``int16``, etc.  The solution for these cases
+would be to rely on in-place operations.
+We currently believe that while less dangerous, this change would affect many users
+and would be surprising more often than not (although expectations differ widely).
+
+In principle, the weak vs. strong behaviour need not be uniform.  It would also
+be possible to make Python floats use the weak behaviour, but Python integers use the
+strong one, since integer overflows are far more surprising.
+
+
+Do not use weak scalar logic in functions
+-----------------------------------------
+
+One alternative to this NEPs proposal is to narrow the use of weak types
+to Python operators.
+
+This has advantages and disadvantages:
+
+* The main advantage is that limiting it to Python operators means that these
+  "weak" types/dtypes are clearly ephemeral to short Python statements.
+* A disadvantage is that ``np.multiply`` and ``*`` are less interchangable.
+* Using "weak" promotion only for operators means that libraries do not have
+  to worry about whether they want to "remember" that an input was a Python
+  scalar initially.  On the other hand, it would add a the need for slightly
+  different (or additional) logic for Python operators.
+  (Technically, probably as a flag to the ufunc dispatching mechanism to toggle
+  the weak logic.)
+* ``__array_ufunc__`` is often used on its own to provide Python operator
+  support for array-likes implementing it.  If operators are special, these
+  array-likes may need a mechanism to match NumPy (e.g. a kwarg to ufuncs to
+  enable weak promotion.)
+
+
+NumPy scalars could be special
+------------------------------
+
+Many users expect that NumPy scalars should be different from NumPy
+arrays, in that ``np.uint8(3) + 3`` should return an ``int64`` (or Python
+integer), when ``uint8_arr + 3`` preserves the ``uint8`` dtype.
+
+This alternative would be very close to the current behaviour for NumPy scalars
+but it would cement a distinction between arrays and scalars (NumPy arrays
+are "stronger" than Python scalars, but NumPy scalars are not).
+
+Such a distinction is very much possible, however, at this time NumPy will
+often (and silently) convert 0-D arrays to scalars.
```

Comment:
To me, it's more important that 0-D arrays behave consistently to their N-D counterparts than that they behave consistently with scalars. Some libraries such as CVXPY have tightened this distinction in recent times, and I think this is a healthy choice.

A numpy 0-D array has a completely different API than a python scalar, and liberally casting one to the other is not beneficial in my opinion. I think they should be _compatible_, but that types should be maintained, e.g. I expect a function operating on scalar values to maintain the input type (0-D array -> 0-D array; python scalar -> python scalar; without silently changing between them).

Reply:
Yes, but NumPy does not currently have that choice, because we create scalars from 0-D arrays silently all the time.  If scalars were distinct objects, you only got with a special operation.  I might agree.

This NEP is only about the *dtype* of the array or scalar, not about this behaviour.  I agree that it should be cleaned up, but it is another set of worms.  But I also think that NumPy scalars have a `dtype` and because of that should behave the same as NumPy arrays for promotion.